### PR TITLE
Webpack updates

### PR DESCRIPTION
**Pull request by gretzky**
Rather than a huge ambitious webpack overhaul, I tackled just 1 big piece and 1 small: the DLL plugin and polyfills.

**DLL**

I've completely removed the DLL setup we had going and replaced it with `webpack.optimization` and `compression-webpack-plugin`.

I've set optimizations for splitting chunks in prod to handle vendor cache groups, which is pretty snappy in development. I've also setup `uglifyjs-webpack-plugin` to enable caching and multi-process parallel running at the same time (this is handled in optimization as well). Along with this, I've added compression plugin to build gzipped bundles. These are served by express and the bundles are cut down by half or more.

**Polyfill**

I've added [`react-app-polyfill`](https://github.com/facebook/create-react-app/blob/master/packages/react-app-polyfill) that has common polyfills for the browser and supports up to IE11.

### CODE REVIEW

**Comment by julienben** [comment]
Wowowow super cool! Great job @gretzky! Very nice to see the ugly config file gone. Will happily review this in-depth on Sunday at the latest unless someone else gets to it first.

**Comment by julienben** [comment]
This is missing some doc updates, no?

**Comment by justingreenberg** [comment]
awesome @gretzky :) i'll also review in depth over the weekend... curious to benchmark initial build/rebuilds after removing DLL configuration

**Comment by julienben** [comment]
Same. I use CSS modules (haven't boarded the CSS-in-JS train yet) and would like to make sure this doesn't dramatically raise update times.

**Comment by DJTB** [comment]
Is `react-app-polyfill` just to slim the dev build?

File: internals/webpack/webpack.base.babel.js [hunk 1]
```diff
@@ -114,11 +114,6 @@ module.exports = options => ({
     ],
   },
   plugins: options.plugins.concat([
-    new webpack.ProvidePlugin({
-      // make fetch available
-      fetch: 'exports-loader?self.fetch!whatwg-fetch',
-    }),
```
[the remark this answers is not in the document]
Comment (gretzky) [reply internals/webpack/webpack.base.babel.js]: Yes, will do. Sorry-- traveling atm but I can tackle these tmrw in airport

**Comment by michaelcuneo** [comment]
Nice. I was just about to implement some compression myself. Now I can just follow this.

File: internals/webpack/webpack.prod.babel.js [hunk 2]
```diff
@@ -19,10 +21,48 @@ module.exports = require('./webpack.base.babel')({
...
+          parse: {},
+          mangle: true,
+          output: {
+            comments: false,
+            ascii_only: true,
+          },
+        },
+        parallel: true,
+        cache: true,
+        sourceMap: true,
```
[the remark this answers is not in the document]
Comment (gretzky) [reply internals/webpack/webpack.prod.babel.js]: Here it just maps error messages to modules

File: package.json [hunk 3]
```diff
@@ -134,6 +114,7 @@
     "babel-plugin-transform-react-remove-prop-types": "0.4.18",
     "circular-dependency-plugin": "5.0.2",
     "compare-versions": "3.4.0",
+    "compression-webpack-plugin": "^2.0.0",
```
[the remark this answers is not in the document]
Comment (gretzky) [reply package.json]: All set!

**Comment by gretzky** [comment]
Doc updates are all set (I think, double check me?)

Otherwise, I introduced gzipping to keep the build size down in prod. The reason why you don't see anything happening in dev is because the plugin is currently only on the production webpack config-- that's why it works on `start:prod`. We could put it in dev, but more than likely it won't cause any significant changes

**Comment by gretzky** [comment]
Looks like there are a lot of failing tests (unrelated to anything in this PR) causing the CI build to fail-- was the dev branch updated and pulled into this PR?

**Comment by julienben** [comment]
Tests were failing because `styled-components` was accidentally restored to v3. It should work now ;)

**Comment by julienben** [comment]
I removed the `whatwg-fetch` import from `request.js` as, if I'm not mistaken, it's made redundant by the presence of `react-app-polyfill/ie11` in the Webpack entry point. Is that correct???

As a consequence, tests for `request.js` were failing so I added `whatwg-fetch` to `request.test.js` only and now `whatwg-fetch` is in dev dependencies.

Let me know if all of this makes sense?

**Comment by gretzky** [comment]
Makes sense to me! Thanks for catching that

**Comment by michaelcuneo** [comment]
I've been following through this, and updating my current project as necessary to adopt these changes because I needed compression in my project... but I notice after Auditing, that it isn't actually loading the .gz files, only the .js version. If I remove the .js version to force Express to load up the .gz's... it does the unexpected token < ... Was compression-webpack-plugin working properly?

**Comment by DJTB** [comment]
Looks like it's working for me - even though it might appear to be loading plain old `.js`.
image

**Comment by michaelcuneo** [comment]
> Looks like it's working for me - even though it might appear to be loading plain old `.js`.
> image

Sorry, my mistake, I was implementing this into my master fork. There are probably a tonne of other implementations into dev that I don't have that are affecting why mine does not work.

**Comment by DJTB** [comment]
@michaelcuneo I should mention that's a production app with the latest RBP changes manually merged in - I haven't actually checked a fresh RBP. 

It's worth cloning RBP master and checking that properly serves the gzip files for you just to be sure.

**Comment by gretzky** [comment]
Yeah, these changes were pulled into the latest version release-- I would try that out. It's a new feature so there might be new kinks to work out! Open to suggestions if it isn't working as expected

**Pull request merged**